src/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,371 @@
+            i: self.i + _rhs.i,
+            j: self.j + _rhs.j,
+            k: self.k + _rhs.k
+        }
+    }
+}
+
+impl std::ops::Sub<Quaternion> for Quaternion {
+    type Output = Quaternion;
+
+    fn sub(self, _rhs: Quaternion) -> Quaternion {
+        Self { 
+            real: self.real - _rhs.real, 
+            i: self.i - _rhs.i,
+            j: self.j - _rhs.j,
+            k: self.k - _rhs.k
+        }
+    }
+}
+
+/// Implements the Hamiltonian product for two quaternions
+impl std::ops::Mul<Quaternion> for Quaternion {
+    type Output = Quaternion;
+
+    fn mul(self, _rhs: Quaternion) -> Self::Output {
+        let real_0 = self.real;
+        let i_0 = self.i;
+        let j_0 = self.j;
+        let k_0 = self.k;
+
+        let real_1 = _rhs.real;
+        let i_1 = _rhs.i;
+        let j_1 = _rhs.j;
+        let k_1 = _rhs.k;
+        
+        let real = real_0*real_1 - i_0*i_1 - j_0*j_1 - k_0*k_1;
+        let i_part = real_0*i_1 + i_0*real_1 + j_0*k_1 - k_0*j_1;
+        let j_part = real_0*j_1 - i_0*k_1 + j_0*real_1 + k_0*i_1;
+        let k_part = real_0*k_1 + i_0*j_1 - j_0*i_1 + k_0*real_1;
+
+        Self::new(real, i_part, j_part, k_part)
+    }
+}
+
+impl ApproxEq for Quaternion {
+    type Margin = F64Margin;
+
+    fn approx_eq<M: Into<Self::Margin>>(self, other: Self, margin: M) -> bool {
+        let margin = margin.into();
+        let diff = other - self;
+        let diff_norm2 = (
+            diff.real * diff.real + diff.i * diff.i + diff.j * diff.j + diff.k * diff.k
+        ).sqrt();
+        diff_norm2.approx_eq(0.0, margin)
+    }
+}
+
+impl PartialEq for Quaternion {
+    fn eq(&self, other: &Self) -> bool {
+        (self.real == other.real) && (self.i == other.i) && (self.j == other.j) && (self.k == other.k)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use float_cmp::{approx_eq};
+
+    use crate::spatialmath::vector3::Vector3;
+    use crate::spatialmath::quaternion::Quaternion;
+
+    #[test]
+    fn new_initializes_quaternion_successfully() {
+        let quat = Quaternion::new(1.0, 0.0, 0.5, 1.0);
+        assert_eq!(quat.real, 1.0);
+        assert_eq!(quat.i, 0.0);
+        assert_eq!(quat.j, 0.5);
+        assert_eq!(quat.k, 1.0)
+    }
+
+    #[test]
+    fn imag_returns_imaginary_components_as_vector() {
+        let quat = Quaternion::new(1.0, 0.0, 0.5, 1.0);
+        let expected_imag = Vector3::new(0.0, 0.5, 1.0);
+        assert_eq!(quat.imag(), expected_imag);
+    }
+
+    #[test]
+    fn set_imag_from_vector_works() {
+        let imag = Vector3::new(0.0, 0.5, 1.0);
+        let expected_quat = Quaternion::new(1.0, 0.0, 0.5, 1.0);
+        let quat = Quaternion::new_from_vector(1.0, imag);
+        assert_eq!(quat, expected_quat);
+    }
+
+    #[test]
+    fn quaternion_normalizes_successfully() {
+        let mut quat = Quaternion::new(
+            0.0,
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * -0.5, 
```

Comment:
I'm mildly surprised that we're multiplying on the right by a constant, and a negative one at that. Could these lines become `-0.5 * (1.0_f64 / 3.0_f64).sqrt())` and similar? Maybe that's too nitpicky and this is good as-is.